r/src/expression.cpp

```diff
@@ -20,82 +20,72 @@
 #if defined(ARROW_R_WITH_ARROW)
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::FieldExpression> dataset___expr__field_ref(
-    std::string name) {
-  return std::make_shared<arrow::dataset::FieldExpression>(std::move(name));
+ds::ExpressionPtr dataset___expr__field_ref(std::string name) {
```

Comment:
Why are these changed to have the parent class as the return value? Just makes for some mismatch because the R classes still use the subclass. Can we use `Expression` everywhere and reduce that boilerplate?